haystack/pipelines/config.py

```diff
@@ -238,7 +294,166 @@ def validate_config(pipeline_config: Dict, strict_version_check: bool = False) -
                 f"Validation failed. {validation.message}. {error_location} " "See the stacktrace for more information."
             ) from validation
 
-    logging.debug(f"Pipeline configuration is valid.")
+    logging.debug(f"The given configuration is valid according to the JSON schema.")
+
+
+def validate_pipeline_graph(pipeline_definition: Dict[str, Any], component_definitions: Dict[str, Any]):
+    """
+    Validates a pipeline's graph without loading the nodes.
+
+    :param pipeline_definition: from get_pipeline_definition()
+    :param component_definitions: from get_component_definitions()
+    """
+    graph = nx.DiGraph()
+    root_node_name = None
+    for node in pipeline_definition["nodes"]:
+        graph, root_node_name = _add_node_to_pipeline_graph(
+            graph=graph, root_node_name=root_node_name, node=node, components=component_definitions
+        )
+    logging.debug(f"The graph for pipeline '{pipeline_definition['name']}' is valid.")
+
+
+def _add_node_to_pipeline_graph(
+    graph: nx.DiGraph,
+    root_node_name: Optional[str],
+    components: Dict[str, Dict[str, str]],
+    node: Dict[str, Any],
+    instance: BaseComponent = None,
+):
+    """
+    Adds a single node to the provided graph, performing all necessary validation steps.
+
+    :param graph: the graph to add the node to
+    :param components: the whole list from get_component_definitions()
+    :param node: `{"name": node_name, "inputs": [node_inputs]}` (the entry to add from get_component_definitions())
+    :param instance: Optional instance of the node. Note that the instance is optional because in some cases
+                     we want to be able to validate the graph without loading the nodes in the process.
+    """
+    # Validate node definition
+    if not instance:
+        _get_class_for_valid_node(node_name=node["name"], components=components)
```

Comment:
While implementing this check I remembered why I did it this way.

The situation is:

- When the instance is not given, that's because we're validating a YAML and we simply need to make sure that there is a valid node definition in the components list, so the node could be instantiated when needed

- If the instance is given, that's because we called `Pipeline.add_node()`, so the node definition will not be in the components list yet and `_get_class_for_valid_node` will fail. 
  - In this case I could surely compute the configuration for the new node and add it to the components list before this check, but it seems redundant, because `_get_class_for_valid_node` really is there only to check if the node could be instantiated. If there's already an instance, the check loses meaning.

I will add a comment to clarify this and rename the `_get_class_for_valid_node` method to make it mode clear. If you have a better way to deal with this let me know :slightly_smiling_face:

Reply:
I eventually modified `Pipeline.add_node` to add the node's definition to the component's list prior to calling this function. LMK if it looks better